File: Data Structures & Algorithms/course-schedule/submission-8.java
class Solution {
    HashMap<Integer, Integer> map = new HashMap<>();

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        if(prerequisites.length == 0){
            return true;
        }
        return canFinishRec(numCourses, prerequisites, 0);
    }

    private boolean canFinishRec(int total, int[][] preReq, int curr) {
        boolean res = true;
        int upper = preReq[curr][0];
        int lower = preReq[curr][1];
        map.put(upper, lower);

        if(curr + 1 < preReq.length){
            res = canFinishRec(total, preReq, curr + 1);
        }

        if(res == false){
            return false;
        }

        for(int i = 0; i < map.size(); i++){
            if(map.containsKey(lower)){
                if(map.get(lower) == upper){
                    return false;
                }
                lower = map.get(lower);
            } else{
                return true;
            }

            if(i == map.size() - 1 && map.containsKey(lower)){
                return false;
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/climbing-stairs/submission-8.py
class Solution:
    def climbStairs(self, n: int) -> int:
        if n == 0 or n == 1 or n == 2 or n == 3:
            return n
        
        i = 4
        res = [3, 2, 1]
        while i <= n:
            res[2] = res[1]
            res[1] = res[0]
            res[0] = res[1] + res[2]
            i += 1

        return res[0]